File: frontend/src/app/services/user/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { RegisterRequest } from '../../models/request/register-request';
import { UserRequest } from '../../models/request/user-request';
import { ChangePasswordRequest } from '../../models/request/change-password-request';
import { ChangeUsernameRequest } from '../../models/request/change-username-request';
import { UserResponse } from '../../models/response/user-response';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root'
})
export class UserService {

    private API_URL = 'http://localhost:8080/api/v1/users';
    constructor(
        private http: HttpClient
    ) {

    }
    createUser(registerRequest: RegisterRequest): Observable<void> {
        return this.http.post<void>(`${this.API_URL}`, registerRequest);
    }

    updateUser(id: string, userRequest: UserRequest): Observable<void> {
        return this.http.put<void>(`${this.API_URL}/${id}`, userRequest);
    }

    changePassword(id: string, changePasswordRequest: ChangePasswordRequest): Observable<void> {
        return this.http.put<void>(`${this.API_URL}/${id}/change-password`, changePasswordRequest);
    }

    changeUsername(id: string, changeUsernameRequest: ChangeUsernameRequest): Observable<void> {
        return this.http.put<void>(`${this.API_URL}/${id}/change-username`, changeUsernameRequest);
    }

    deleteUser(id: string): Observable<void> {
        return this.http.delete<void>(`${this.API_URL}/${id}`);
    }
    
    getUser(id: string): Observable<UserResponse> {
        return this.http.get<UserResponse>(`${this.API_URL}/${id}`);
    }

    getAllUsers(): Observable<UserResponse[]> {
        return this.http.get<UserResponse[]>(`${this.API_URL}`);
    }

}

File: backend/src/main/java/org/gestionstock/stock/Service/UserService.java
package org.gestionstock.stock.Service;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import org.gestionstock.stock.Entity.User;
import org.gestionstock.stock.EntityRepository.UserRepository;
import org.gestionstock.stock.Exception.InvalidCredentialsException;
import org.gestionstock.stock.Exception.UserNotFoundException;
import org.gestionstock.stock.Exception.UsernameAlreadyExistException;
import org.gestionstock.stock.IService.IUserService;
import org.gestionstock.stock.Payload.Mapper.UserMapper;
import org.gestionstock.stock.Payload.Request.ChangePasswordRequest;
import org.gestionstock.stock.Payload.Request.ChangeUsernameRequest;
import org.gestionstock.stock.Payload.Request.RegisterRequest;
import org.gestionstock.stock.Payload.Request.UserRequest;
import org.gestionstock.stock.Payload.Response.UserResponse;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class UserService implements IUserService{
    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;
    
    @Override
    @Transactional
    public void createUser(RegisterRequest RegusterRequest) {
        User user = userMapper.toUser(RegusterRequest);
        log.info("User: {}", user);
        user.setPassword(passwordEncoder.encode(RegusterRequest.password()));
        userRepository.save(user);
        log.info("User saved");
    }

    @Override
    @Transactional
    public void updateUser(String id, UserRequest userRequest) {
        User user = userRepository.findById(UUID.fromString(id))
            .orElseThrow(() -> new UserNotFoundException());
        log.info("User: {}", user);
        user.setFirstName(userRequest.firstName());
        user.setLastName(userRequest.lastName());
        user.setAddress(userRequest.address());
        user.setPhone(userRequest.phone());
        user.setImageUrl(userRequest.imageUrl());
        user.setRoles(userRequest.roles());
        userRepository.save(user);
        log.info("User updated");
    }

    @Override
    @Transactional
    public void changePassword(String id, ChangePasswordRequest changePasswordRequest) {
        User user = userRepository.findById(UUID.fromString(id))
            .orElseThrow(() -> new UserNotFoundException());
        log.info("User: {}", user);
        if(!passwordEncoder.matches(changePasswordRequest.oldPassword(), user.getPassword())){
            log.info("Invalid Password Request");
            throw new InvalidCredentialsException();
        }
        user.setPassword(passwordEncoder.encode(changePasswordRequest.newPassword()));
        userRepository.save(user);
        log.info("Password changed");
    }

    @Override
    @Transactional
    public void changeUsername(String id, ChangeUsernameRequest changeUsernameRequest) {
        User user = userRepository.findById(UUID.fromString(id))
            .orElseThrow(() -> new UserNotFoundException());
        log.info("User: {}", user);
        if(userRepository.existsByUsername(changeUsernameRequest.newUsername())){
            log.info("Username already exist");
            throw new UsernameAlreadyExistException();
        }
        user.setUsername(changeUsernameRequest.newUsername());
        userRepository.save(user);
        log.info("Username changed");
    }

    @Override
    @Transactional
    public void deleteUser(String id) {
        if(!userRepository.existsById(UUID.fromString(id))){
            log.info("User not found");
            throw new UserNotFoundException();
        }
        userRepository.deleteById(UUID.fromString(id));
        log.info("User deleted");
    }

    @Override
    public UserResponse getUser(String id) {
        log.info("User id: {}", id);
        return userRepository.findById(UUID.fromString(id))
            .map(userMapper::fromUser)
            .orElseThrow(() -> new UserNotFoundException());
    }

    @Override
    public List<UserResponse> getAllUsers() {
        log.info("All users");
        return userRepository.findAll()
            .stream()
            .map(userMapper::fromUser)
            .collect(Collectors.toList());
    }
    
}
